File: 1315-sum-of-nodes-with-even-valued-grandparent/1315-sum-of-nodes-with-even-valued-grandparent.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumEvenGrandparent(self, root: TreeNode) -> int:
        
        def dfs(cur: TreeNode, parent: TreeNode, grandparent: TreeNode) -> int:
            if not cur:
                return 0
            cur_sum = 0
            if grandparent and grandparent.val % 2 == 0:
                cur_sum += cur.val
            return cur_sum + dfs(cur.left, cur, parent) + dfs(cur.right, cur, parent)
        
        return dfs(root, None, None)

File: 1315-sum-of-nodes-with-even-valued-grandparent/1315-sum-of-nodes-with-even-valued-grandparent.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun dfs(cur: TreeNode?, parent: TreeNode?, grandparent: TreeNode?): Int {
        if (cur === null) return 0
        var sum = 0
        if (grandparent !== null && grandparent.`val` % 2 === 0) sum += cur.`val`
        return sum + dfs(cur.left, cur, parent) + dfs(cur.right, cur, parent)
    }
    
    fun sumEvenGrandparent(root: TreeNode?): Int {
        return dfs(root, null, null)
    }
}